src/bin/tikv-ctl.rs

```diff
@@ -182,3 +304,122 @@ fn dump_range(db: DB, from: String, to: Option<String>, limit: Option<u64>, cf:
                  })
         .unwrap()
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::str;
+    use rocksdb::Writable;
+    use tikv::util::codec::bytes::encode_bytes;
+    use tikv::util::codec::number::NumberEncoder;
+    use tikv::raftstore::store::keys;
+    use tikv::storage::{ALL_CFS, CF_LOCK, CF_WRITE, CF_DEFAULT};
+    use tikv::storage::mvcc::{Lock, Write, LockType, WriteType};
+    use tempdir::TempDir;
+    use tikv::util::rocksdb::new_engine;
+    use tikv::storage::types::Key;
+
+    const PREFIX: &'static [u8] = b"k";
+
+    #[test]
+    fn test_ctl_mvcc() {
+        let tmp_dir = TempDir::new("mvcc_tmp").expect("create mvcc_tmp dir");
+        let file_path = tmp_dir.path().join("tmp_db");
+        let db = new_engine(file_path.to_str().unwrap(), ALL_CFS).unwrap();
+        let test_data = vec![(PREFIX, b"v", 5), (PREFIX, b"x", 10), (PREFIX, b"y", 15)];
+        let keys: Vec<_> = test_data.iter()
+            .map(|data| {
+                keys::data_key(Key::from_raw(&data.0[..]).append_ts(data.2).encoded().as_slice())
+            })
+            .collect();
+        let default_value: Vec<_> = test_data.iter().map(|data| *data.1).collect();
+        let kvs = keys.iter().zip(default_value.iter());
+        for (k, v) in kvs {
+            db.put(k.as_slice(), v).unwrap();
+        }
+        let kvs_gen: Vec<MvccKv<Vec<u8>>> = gen_mvcc_iter(&db, "k", CF_DEFAULT).unwrap();
+        let mut test_iter = test_data.clone();
+        assert_eq!(test_iter.len(), test_data.len());
+        for kv in kvs_gen {
+            let ts = kv.key.decode_ts().unwrap();
+            let key = kv.key.truncate_ts().unwrap().raw().unwrap();
+            let value = kv.value.as_slice();
+            test_iter.retain(|&s| !(&s.0[..] == key.as_slice() && &s.1[..] == value && s.2 == ts));
+        }
+        assert_eq!(test_iter.len(), 0);
+
+        // Test MVCC Lock
+        let test_data_lock = vec![(b"kv", LockType::Put, b"v", 5),
+                                  (b"kx", LockType::Lock, b"x", 10),
+                                  (b"kz", LockType::Delete, b"y", 15)];
+        let keys: Vec<_> = test_data_lock.iter()
+            .map(|data| {
+                let encoded = encode_bytes(&data.0[..]);
+                keys::data_key(encoded.as_slice())
+            })
+            .collect();
+        let lock_value: Vec<_> = test_data_lock.iter()
+            .map(|data| Lock::new(data.1, data.2.to_vec(), data.3).to_bytes())
+            .collect();
+        let kvs = keys.iter().zip(lock_value.iter());
+        println!("KVs: {:?}", kvs.clone().collect::<Vec<_>>());
+        let lock_cf = db.cf_handle(CF_LOCK).unwrap();
+        for (k, v) in kvs {
+            db.put_cf(lock_cf, k.as_slice(), v.as_slice()).unwrap();
+        }
+        let kvs_gen: Vec<MvccKv<Lock>> = gen_mvcc_iter(&db, "k", CF_LOCK).unwrap();
```

Comment:
Here we should use 'kv', 'kx', 'kz' to dump data separately, and each scan should only return 1 result.